README.md
# gnome-hearts
The classic Hearts game for Gnome


ChangeLog
gnome-hearts (0.3-2.2) unstable; urgency=medium

  * Non-maintainer upload.
  * Don't recommend gnome-games-extra-data. This package is obsolete and
    doesn't provide anything gnome-hearts could use anyway.
  * Build using dh-python instead of python-support.  Closes: #786141

 -- Mattia Rizzolo <mattia@debian.org>  Sun, 13 Dec 2015 16:20:50 +0000

gnome-hearts (0.3-2.1) unstable; urgency=low

  * Non-maintainer upload.
  * Drop gnome-cards patch to use upstream card
    configuration. (Closes: #660668)

 -- Neil Williams <codehelp@debian.org>  Sun, 08 Jul 2012 16:42:25 -0600

gnome-hearts (0.3-2) unstable; urgency=low

  * Build-depend on intltool to fix FTBFS.

 -- Josselin Mouette <joss@debian.org>  Mon, 28 Jul 2008 12:01:15 +0200

gnome-hearts (0.3-1) unstable; urgency=low

  [ Sander Marechal ]
  * New upstream release
   - Three new AI opponents
   - Bugfixes in the existing AI and scoring
   - Transation updates

  [ Josselin Mouette ]
  * Remove versioned build-dependency on scrollkeeper. Closes: #490830.

 -- Sander Marechal <s.marechal@jejik.com>  Thu, 24 Jul 2008 21:52:33 +0200

gnome-hearts (0.2.1-1) unstable; urgency=low

  [ Sander Marechal ]
  * New upstream release
    - 01_gnome_cards_data.patch: path to gnome-cards-data cards fixed upstream
    - 02_fix_pythondir.patch is fixed upstream
    - 03_shared_link.patch is fixed upstream
    - 04_fix_desktop.patch is fixed upstream

  [ Sebastian Dröge ]
  * Update Standards-Version to 3.7.3, no additional changes needed.

 -- Sebastian Dröge <slomo@debian.org>  Tue, 11 Mar 2008 14:51:44 +0100

gnome-hearts (0.2-1) unstable; urgency=low

  [ Sander Marechal ]
  * New upstream release. Closes: #425670, #420861.
    - Big rewrite of the C code
    - Uses Python instead of Lua for AI scripts
    - Many bugfixes and translation updates
  * Dependency on Lua replaced with Python
  * Recreated 01_gnome_cards_data.patch for the new upstream source

  [ Josselin Mouette ]
  * 02_fix_pythondir.patch: install the private python modules in a
    private directory.
  * Call dh_pysupport; build-depend on python-support 0.4.
  * Don't hardcode the 2.4 version of python.
  * 03_shared_link.patch: link to python dynamically.
  * 01_gnome_cards_data.patch:
    + Fix the path to the cards in gnome-cards-data.
    + Don't remove the picture, it's useless to do it in the patch.

 -- Josselin Mouette <joss@debian.org>  Tue, 23 Oct 2007 11:49:44 +0200

gnome-hearts (0.1.3-4) unstable; urgency=low

  * 01_gnome_cards_data.patch: use the new path for cards.
  * Depend on gnome-cards-data 2.20.

 -- Josselin Mouette <joss@debian.org>  Sat, 29 Sep 2007 03:40:20 +0200

gnome-hearts (0.1.3-3) unstable; urgency=high

  * Add a get-orig-source target to retrieve the upstream tarball.
  * Depend on librsvg2-common for SVG support; closes: #421372.
  * Wrap build-deps and deps.
  * Urgency high for RC bugfix.

 -- Loic Minier <lool@dooz.org>  Mon, 30 Apr 2007 13:52:21 +0200

gnome-hearts (0.1.3-2) unstable; urgency=high

  [ Sander Marechal ]
  * Fixed the 01_gnome_cards_data.patch. Closes: #402715

  [ Loic Minier ]
  * Urgency high for RC fix.

 -- Loic Minier <lool@dooz.org>  Wed, 13 Dec 2006 00:29:47 +0100

gnome-hearts (0.1.3-1) unstable; urgency=low

  * New upstream release. Closes: #396043.
    - Fixes crashes when card style or background cannot be loaded
    - Updated the translations
  * Backed out the bonded.png style again. The new upstream release will
    not crash when a style cannot be found but load another available
    card style instead. This removes the sng dependency and the conflict
    with gnome-cards-data << 1:2.16

 -- Sander Marechal <s.marechal@jejik.com>  Sat, 11 Nov 2006 18:05:00 +0100

gnome-hearts (0.1.2-2) unstable; urgency=high

  * Add CDBS' utils.
  * Update patch 01_gnome_cards_data to use bonded.svg as the default card set
    instead of bonded.png; depend on gnome-cards-data >= 1:2.16; partially
    fixes #396043.
  * Extract bonded.png from older gnome-cards-data and install it where
    expected for backwards compatibility with end-user configs; build-depend
    on sng; conflict with gnome-cards-data << 1:2.16; partially fixes #396043.

 -- Loic Minier <lool@dooz.org>  Mon, 30 Oct 2006 09:58:17 +0100

gnome-hearts (0.1.2-1) unstable; urgency=low

  * New upstream release:
    - Fixes a crash in the Luis AI
    - Updated the translations
  * Add watch file.
  * Update copyright URL.

 -- Sander Marechal <s.marechal@jejik.com>>  Thu,  7 Sep 2006 00:54:19 +0200

gnome-hearts (0.1.1-1) unstable; urgency=low

  [ Sander Marechal ]
  * Initial release. Closes: #370036,

  [ Josselin Mouette ]
  * Move the executable to /usr/games.

 -- Josselin Mouette <joss@debian.org>  Mon,  3 Jul 2006 19:50:22 +0200

2008-07-24  sander

	* [r206] /branches/0.3.x: Branching version 0.3 for release
	* [r205] /trunk/NEWS: Updated NEWS
	* [r204] /trunk/man/gnome-hearts.6: Update man page
	* [r203] /trunk/help/C/gnome-hearts.xml,
	  /trunk/help/nl/gnome-hearts.xml, /trunk/po/nl.po: Updated the
	  documentation
	* [r202] /trunk/src/gnome-hearts.glade: Update Glade file to new
	  glade version
	* [r201] /trunk/configure.in, /trunk/po/da.po, /trunk/po/de.po,
	  /trunk/po/el.po, /trunk/po/en_GB.po, /trunk/po/es.po,
	  /trunk/po/eu.po, /trunk/po/fi.po, /trunk/po/fr.po,
	  /trunk/po/hu.po, /trunk/po/id.po, /trunk/po/it.po,
	  /trunk/po/ml.po, /trunk/po/nl.po, /trunk/po/pl.po,
	  /trunk/po/pt.po, /trunk/po/pt_BR.po, /trunk/po/ru.po,
	  /trunk/po/sk.po, /trunk/po/sv.po, /trunk/po/tl.po,
	  /trunk/po/zh_CN.po, /trunk/src/gnome-hearts.glade: Bump version
	  number to 0.3
	* [r200] /trunk/AUTHORS, /trunk/scripts/players/Makefile.am,
	  /trunk/scripts/players/peter.py: Added the Peter AI. Thanks go to
	  Charles Crayne
	* [r199] /trunk/src/cfg.c, /trunk/src/cfg.h, /trunk/src/hearts.c:
	  New rules should be published when a new game starts, which is
	  before a configuration change is saved

2008-07-23  sander


scripts/hearts.py
# Hearts - hearts.py
# Copyright 2006 Sander Marechal
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

# This is an attempted translation of hearts.lua
# Cards are defined as tuples, so card = (suit, rank)
# Multiple cards are passed around as lists of tuples.

import os
import sys
import string

sys.path.append("/usr/share/gnome-hearts");
sys.path.append("/usr/share/gnome-hearts/players");

from definitions import *
from stock_ai import *

# Load all the available player AI's
for script in os.listdir(os.path.abspath("/usr/share/gnome-hearts/players")):
	module_name, ext = os.path.splitext(script)
	if ext == '.py':
		module = __import__(module_name)
		for method in dir(module):
			if method.startswith('PlayerAI'):
				# Make the AI class a global
				globals()[method] = getattr(module, method)
				# Register the AI in C
				ai = globals()[method](None, None, None, None)
				register_ai(ai.name, method)

# The Trick class
class Trick:
	"""A trick of cards, equal to the C version"""
	
	def __init__(self):
		self.card = []
		self.trump = None
		self.num_played = 0
		self.first_played = None
	
	def set(self, list, trump, num_played, first_played):
		self.card = list
		self.trump = trump
		self.num_played = num_played
		self.first_played = first_played
	
	def winner(self):
		"""Return who played the highest card on the trick"""
		if self.num_played > 0:
			return self.card.index(self.get_highest_card())
		return None
	
	def get_highest_card(self):
		"""Return the highest trump card in the list"""
		highest_card = (0, 0)
		for card in self.card:
			if card != None and card[0] == self.trump and card[1] > highest_card[1]:
				highest_card = card
		return highest_card

trick = Trick()

# The score class
class Score:
	"""This object contains the score of this round and the entire game so far"""

	def __init__(self):
		self.round = [0, 0, 0, 0]
		self.game = [0, 0, 0, 0]
	
	def set(self, round, game):
		self.round = round
		self.game = game

score = Score()

# The rules class
class Rules:
	"""This object contains all the rules for the game"""

	def __init__(self):
		self.ruleset = "standard"
		self.clubs_lead = True
		self.hearts_break = True
		self.queen_breaks_hearts = False
		self.no_blood = True

	def set(self, ruleset, clubs_lead, hearts_break, queen_breaks_hearts, no_blood):
		self.ruleset = ruleset
		self.clubs_lead = clubs_lead
		self.hearts_break = hearts_break
		self.queen_breaks_hearts = queen_breaks_hearts
		self.no_blood = no_blood

rules = Rules()

# General setting functions used by C
def player_set(object, direction):
	player[direction] = object

# The list of players
player = [None, None, None, None]



src/hearts.c
/*
 *  Hearts - hearts.c
 *  Copyright 2006 Sander Marechal
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "hearts.h"
#include "background.h"

Trick 		game_trick;
gint 		game_state;
gint 		game_pass;
gint 		game_trick_winner;
gboolean 	game_hearts_broken;
CardsDeck 	*game_deck;
GList	 	*game_score_labels;

Player *player[4]; 	/* the players of the game */
gint user;			/* which playes is the user */

/* Definition of the commandline options */
static GOptionEntry option_entries[] =
{
	{"show-cards", 0, 0, G_OPTION_ARG_NONE, &cfg.show_cards, N_("Show all the card faces while playing"), NULL},
	{NULL}
};

/* Load a pixmap */
GdkPixmap* get_pixmap(const char *filename)
{
	GdkPixmap *ret;
	GdkPixbuf *im;
	
	if (filename == NULL)
		return NULL; 
	
	im = gdk_pixbuf_new_from_file (filename, NULL);
	if (im != NULL)
	{
		gdk_pixbuf_render_pixmap_and_mask_for_colormap (im, gdk_colormap_get_system(), &ret, NULL, 127);
		gdk_pixbuf_unref (im);
	}
	else
	{
		ret = NULL;
	}
	
	return ret;
}

/* start a new game */
void game_new(void)
{
	gint i;
	GList *l;
	GtkTable *score_table;
	GtkWidget *widget;
	
	/* reset the score_table. To do this, we need to remove the child widgets and then resize */
	score_table = GTK_TABLE(glade_xml_get_widget(ui.xml, "score_table"));
	for (l = game_score_labels; l; l = l->next)
	{
		GtkLabel *label;
		label = l->data;
		gtk_widget_destroy(GTK_WIDGET(label));
	}
	g_list_free(game_score_labels);
	game_score_labels = NULL;
	gtk_table_resize(score_table, 2, 4);
	
	/* set the score window title */
	gtk_window_set_title(GTK_WINDOW(glade_xml_get_widget(ui.xml, "score")), _("Scores"));
	
	/* Publish the rulesets to Python. This is done here because game_new() is called when the
	 * rules have changed, and player_new() below initializes the players for the new game.
	 */
	config_publish();

	/* refresh the AI's */
	for (i = 0; i < 4; i++)
		player_free(player[i]);
	
	widget = glade_xml_get_widget(ui.xml, "playingarea");

	gint facing = cfg.show_cards ? FACE_UP : FACE_DOWN;
	player[0] = player_new(NORTH, facing, cfg.north_ai, &game_trick, widget);
	player[1] = player_new(EAST, facing, cfg.east_ai, &game_trick, widget);
	player[2] = player_new(SOUTH, FACE_UP, NULL, &game_trick, widget);
	player[3] = player_new(WEST, facing, cfg.west_ai, &game_trick, widget);
	
	game_pass = 0;
	game_new_round();
}

/* start a new round in this game */
void game_new_round(void)
{
	gint i;
	
	/* reset the game */
	cards_deck_reset(game_deck);
	trick_reset(&game_trick);
	trick_publish(&game_trick);
	game_pass = (game_pass + 1) % 4;
	game_hearts_broken = FALSE;
	
	/* new hands for everyone */
	for (i = 0; i < 4; i++)
	{
		cards_hand_free_cards(player[i]->hand);
		cards_hand_draw(player[i]->hand, game_deck, 13);
		g_list_free(player[i]->cards_taken);
		player[i]->cards_taken = NULL;
		player[i]->score_round = 0;
	}

	game_score_publish();

	/* have the AI's select cards to pass on if we're passing cards this round */
	for (i = 0; i < 4; i++)
	{
		if (i != user && game_pass > 0)
			player_mark_selected(player[i], player_select_cards(player[i]));
	}
	
	/* set the game state */
	if (game_pass > 0)
	{
		game_set_status(_("Select three cards that you wish to pass on."));
		game_state = GAME_SELECT_CARDS;
	}
	else
	{
		game_set_status(_("Click on the card you wish to play."));
		game_state = GAME_PLAY;
		game_open_round();
	}
}

/* restart the current round */
void game_restart_round(void)
{
	gint i;
	
	/* reset the game */
	trick_reset(&game_trick);
	trick_publish(&game_trick);
	game_hearts_broken = FALSE;
	
	/* reset the hands for everyone */
	for (i = 0; i < 4; i++)
	{
		cards_hand_reset(player[i]->hand);
		g_list_free(player[i]->cards_taken);
		player[i]->cards_taken = NULL;
		player[i]->score_round = 0;
	}
	
	game_score_publish();

	/* have the AI's select cards to pass on if we're passing cards this round */
	for (i = 0; i < 4; i++)
	{
		if (i != user && game_pass > 0)
			player_mark_selected(player[i], player_select_cards(player[i]));
	}
	
	/* set the game state */
	if (game_pass > 0)
	{
		game_set_status(_("Select three cards that you wish to pass on."));
		game_state = GAME_SELECT_CARDS;
	}
	else
	{
		game_set_status(_("Click on the card you wish to play."));
		game_state = GAME_PLAY;
		game_open_round();
	}
}

/* start the round and play untill the player is due */
void game_open_round (void)
{
	gint i, starter = user;
	Card *two_of_clubs;
	
	/* figure out which playes has to start */
	if (cfg.clubs_lead)
	{
		two_of_clubs = &g_array_index(game_deck, Card, CARD_ID(CARDS_CLUBS, CARD_TWO));
		for (i = 0; i < 4; i++)
		{
			if (g_list_find(player[i]->hand->list, two_of_clubs))
				starter = i;
		}
	}
	else
	{
		starter = (game_pass + 1) % 4;
	}
	
	/* play */
	i = starter;
	while (i != user)
	{
		player_play_card(player[i], player_play(player[i]));
		i = (i + 1) % 4;
	}
}

/* test for the end of the round and show this game's score */
gboolean game_end_test(void)
{
	gint rows, i, score_value = 0;
	GtkWidget *label, *score;
	GtkTable *score_table;
	gchar *label_text;
	GList *l;
	
	/* end of round test */
	if (g_list_length(player[user]->hand->list) == 0)
	{
		/* Signal the players that the round is over */
		player_round_end(player[NORTH]);
		player_round_end(player[EAST]);
		player_round_end(player[WEST]);
		
		/* strike out the previous row */
		i = 0;
		l = game_score_labels;
		while (l != NULL && i < 4)
		{
			GtkLabel *label = l->data;
			gchar *text = g_strdup_printf("<s>%s</s>", gtk_label_get_label(label));
			gtk_label_set_markup(label, text);
			g_free(text);
			
			l = l->next;
			i++;
		}
		
		/* resize the score_table widget */
		score_table = (GtkTable*)glade_xml_get_widget(ui.xml, "score_table");
		g_object_get(score_table, "n-rows", &rows, NULL);
		gtk_table_resize(score_table, rows + 1, 4);
		
		/* fill the new row */
		for (i = 0; i < 4; i++)
		{
			/* update the scores */
			player[i]->score_total += player[i]->score_round;
			player[i]->score_round = 0;
			
			label_text = g_strdup_printf("%d", player[i]->score_total);
			label = gtk_label_new(label_text);
			game_score_labels = g_list_prepend(game_score_labels, label);
			g_free(label_text);
				
			gtk_label_set_justify((GtkLabel*)label, GTK_JUSTIFY_CENTER);
			gtk_table_attach(score_table, label, i, i + 1, rows, rows + 1,
					 GTK_EXPAND, GTK_SHRINK, 2, 0);
		}
		
		/* grab the score widget */		
		score = glade_xml_get_widget(ui.xml, "score");
		
		/* end of game test */
		for (i = 0; i < 4; i++)
			score_value = MAX(score_value, player[i]->score_total);
		
		if (score_value >= (cfg.ruleset == RULESET_SPOT_HEARTS ? 500 : 100))
		{
			game_state = GAME_END;
			
			gint winner = NORTH;
			for (i = 0; i < 4; i++)
			{
				if (player[i]->score_total < score_value)
				{
					score_value = player[i]->score_total;
					winner = i;
				}
			}
			
			if (winner == user)
				gtk_window_set_title((GtkWindow*)score, _("Game over - You have won!"));
			else
				gtk_window_set_title((GtkWindow*)score, _("Game over - You have lost"));
			
			game_set_status(_("Click somewhere to start a new game."));
		}
		else
		{
			game_set_status(_("Click somewhere to continue the game."));
			game_state = GAME_ROUND_END;
		}
		
		/* show the score */
		gtk_widget_show_all(score);
		
		return TRUE;
	}
	
	return FALSE;
}

/* pass cards to the next player */
gboolean game_pass_cards(GtkWidget *widget)
{
	gint x, y, width, height, mouse_x, mouse_y, p;
	GList *temp[4];
	
	/* get the area we're supposed to click on */
	gtk_widget_get_pointer(widget, &mouse_x, &mouse_y);
	cards_hand_get_area(player[(user + game_pass) % 4]->hand, widget, ui.cards_image, &x, &y, &width, &height, TRUE);
	
	if (mouse_x >= x && mouse_x < x + width && mouse_y >= y && mouse_y < y + height)
	{
		/* pass the cards and make a copy in the history for trick reset support */
		for (p = 0; p < 4; p++)
			temp[p] = cards_hand_draw_selected(player[p]->hand);
		
		for (p = 0; p < 4; p++)
		{
			player_receive_cards(player[(p + game_pass) % 4], temp[p]);
			g_list_free(temp[p]);
		}
		
		/* redraw */
		ui_draw_playingarea();
		return TRUE;
	}
	
	return FALSE;
}

/* (de)select cards in the passing round */
gint game_select_cards(GtkWidget *widget)
{
	GList *i;
	Card *card, *active;
	gint x, y, width, height, num_selected = 0;
	
	i = player[user]->hand->list;
	active = NULL;
	
	/* loop through the hand to find the active cards and the number of selected cards */			
	do
	{
		card = (Card*)i->data;
		
		if (card->active)
			active = card;
		
		if (card->selected)
			num_selected++;
							
		i = g_list_next(i);
	} while (i != NULL);
	
	/* You cannot select the active card if there are already 3 selected cards */				
	if (active != NULL && (num_selected < 3 || active->selected == TRUE))
	{
		active->selected = !active->selected;
		
		/* adjust the number of selected cards */
		if (active->selected)
			num_selected++;
		else
			num_selected--;
	}
	
	/* redraw the screen */
	cards_hand_get_area(player[user]->hand, widget, ui.cards_image, &x, &y, &width, &height, TRUE);
	background_render(ui.backbuffer, x, y, width, height);
	cards_hand_render(player[user]->hand, ui.cards_image, ui.backbuffer, ui.backbuffer_gc, x, y);
	gtk_widget_queue_draw_area(widget, x, y, width, height);
	
	return num_selected;
}

/* Publish the game's score to Python */
void game_score_publish()
{
	PyObject *module, *dict, *object, *result;
	
	/* Get a reference to score.set() */
	module = PyImport_AddModule("__main__");
	dict = PyModule_GetDict(module);
	object = PyDict_GetItemString(dict, "score");
	
	PyObject *score_list_round = Py_BuildValue("[iiii]",
		player[NORTH]->score_round,
		player[EAST]->score_round,
		player[SOUTH]->score_round,
		player[WEST]->score_round);
	
	PyObject *score_list_game = Py_BuildValue("[iiii]",
		player[NORTH]->score_total + player[NORTH]->score_round,
		player[EAST]->score_total + player[EAST]->score_round,
		player[SOUTH]->score_total + player[SOUTH]->score_round,
		player[WEST]->score_total + player[WEST]->score_round);

	result = PyObject_CallMethod(object, "set", "OO", score_list_round, score_list_game);

	Py_DECREF(score_list_round);
	Py_DECREF(score_list_game);
	Py_DECREF(result);
}

/* returns wether a card is valid to play from this hand on the stack */
gboolean game_is_valid_card(Card *card, CardsHand *hand, Trick *trick)
{
	/* special case: opening with a two of clubs (variant) */
	if (cfg.clubs_lead && g_list_length(hand->list) == 13 && trick->num_played == 0)
	{
		if (card->suit == CARDS_CLUBS && card->rank == CARD_TWO)
			return TRUE;
		else
			return FALSE;
	}
	
	/* special case, no blood in the first round, unless we have to */
	if (cfg.no_blood
		&& (card->suit == CARDS_HEARTS || (card->suit == CARDS_SPADES && card->rank == CARD_QUEEN))
		&& g_list_length(hand->list) == 13)
	{
		GList *l;
		gint num_pointcards = 0;
		for (l = hand->list; l; l = l->next)
		{
			Card *lcard = l->data;
			if (lcard->suit == CARDS_HEARTS || (lcard->suit == CARDS_SPADES && lcard->rank == CARD_QUEEN))
				num_pointcards++;
		}
		
		if (num_pointcards < 13)
			return FALSE;
	}
	
	/* special case: hearts must have been "broken" to be played as trump (variant).
	If a player has all hearts, then one must be played anyway */
	if (cfg.hearts_break 
		&& trick->num_played == 0
		&& game_hearts_broken == FALSE
		&& card->suit == CARDS_HEARTS
		&& cards_hand_num_suit(hand, CARDS_HEARTS) != cards_hand_length(hand)
	)
		return FALSE;
	
	/* follow the trump if possible */
	if (trick->trump > -1 && card->suit != trick->trump && cards_hand_num_suit(hand, trick->trump) > 0)
		return FALSE;
	
	/* no more reason to disallow a card */
	return TRUE;
}

/* get a hint string */
gchar* game_get_hint(void)
{
	if (game_state == GAME_PLAY)
	{
		if (game_trick.num_played == 4)
			return g_strdup(_("Click somewhere to continue the game."));
		
		Card *card = player_play(player[user]);
		/* Hint: %s is the name of the card the player should play */
		return g_strdup_printf(_("Play %s."), card_get_name(card));
	}
	else if (game_state == GAME_PASS_CARDS)
	{
		gchar *name = player[(user + game_pass) % 4]->name;
		return g_strdup_printf(_("Click on the hand of %s to pass the cards."), name);
	}
	else if (game_state == GAME_SELECT_CARDS)
	{
		GList *l, *cards = player_select_cards(player[user]);
		
		for (l = cards; l; l = l->next)
		{
			Card *card = l->data;
			
			if (!card->selected)
				/* Hint: %s is the name of the card the player should pass on */
				return g_strdup_printf(_("Pass %s."), card_get_name(card));
		}
	}
	
	return g_strdup(_("Sorry, I can't help you."));
}

/* Set the text of the status bar */
void game_set_status(gchar *message)
{
	GtkStatusbar *statusbar = GTK_STATUSBAR(glade_xml_get_widget(ui.xml, "statusbar"));
	gint context_id = gtk_statusbar_get_context_id(statusbar, "hearts user feedback");
	gtk_statusbar_pop(statusbar, context_id);
	gtk_statusbar_push(statusbar, context_id, message);
}

int main(int argc, char *argv[])
{
	#ifdef ENABLE_NLS
		bindtextdomain(GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
		textdomain(PACKAGE);
		setlocale(LC_ALL, "");
	#endif

	/* Initialize GNOME */
	GOptionContext *option_context = g_option_context_new(_("- Play a game of hearts"));
	g_option_context_add_main_entries(option_context, option_entries, GETTEXT_PACKAGE);

	gnome_program_init(PACKAGE, VERSION, LIBGNOMEUI_MODULE,
			argc, argv,
			GNOME_PARAM_GOPTION_CONTEXT, option_context,
			GNOME_PARAM_APP_DATADIR, PACKAGE_DATA_DIR,
			GNOME_PARAM_NONE);
	
	/* init the game's global variables */
	game_score_labels = NULL;
	game_state = GAME_SELECT_CARDS;
	game_pass = 0;
	game_hearts_broken = FALSE;
	game_deck = cards_deck_new(TRUE, FALSE);
	trick_reset(&game_trick);
	user = SOUTH;
	
	/* Start the Glade UI - must be done before starting Python because Python loads AI's in the UI combo boxes */ 
	ui_load();
	
	/* Start the Python interpreter and load the hearts base module and list of AI's */
	python_start();
	
	/* Load the configuration */
	config_load();
	
	/* Apply the loaded configuration settings */
	ui_set_config();
	
	/* Load the background */
	background_load(cfg.background_image);
	
	/* Start the game */
	ui_start();
	game_new();
	gtk_main();
	
	/* Free the global structures */
	background_free();
	config_free();
	python_stop();
	ui_free();

	return 0;
}
